### DISCUSSION TITLE
Headless VersionTracking

### DISCUSSION THREAD

**Discussion by clearbluejar** [NationalSecurityAgency/ghidra#5362] [Q&A] [answered]
Is there any possible way to get [AutoVersionTrackingScript.java](https://github.com/NationalSecurityAgency/ghidra/blob/894bd3cfc2db78686727bbf2a1d369de92878033/Ghidra/Features/VersionTracking/ghidra_scripts/AutoVersionTrackingScript.java) to run without the need for a "tool"? I see in the script that it uses a tool to create a VTPlugin, but there seems to be no ability to create it in headless.  I see that this has been brought up before https://github.com/NationalSecurityAgency/ghidra/issues/934

I am attempting to automate several runs of the version tracking tool simply using with the automated version tracking task.  Is there a way?

Some things I don't quite understand. What is a "tool"?  Is there a way to fake it enough to run the script?
https://github.com/NationalSecurityAgency/ghidra/blob/894bd3cfc2db78686727bbf2a1d369de92878033/Ghidra/Features/VersionTracking/ghidra_scripts/AutoVersionTrackingScript.java#L65-L70

I tried to fake it, and create my own `GhidraTool` instance, so that I could add a plugin to it, but it complained about not having X11.  

![image](https://github.com/NationalSecurityAgency/ghidra/assets/3752074/00c83492-d608-4787-aa5b-6ef16321b4a1)

Then I added the X11 settings for CI  from https://github.com/NationalSecurityAgency/ghidra/blob/master/DevGuide.md#setup-build-in-ci

Then I received a runtime error about:

![image](https://github.com/NationalSecurityAgency/ghidra/assets/3752074/8d369aef-cca2-4eea-9a70-12d28757943d)

Perhaps I could call my code from `SwingUtilities.invokeAndWait`? I'm really just stumbling around in the dark on this one.

**Comment by astrelsky** [comment]
A `Tool` is the `CodeBrowser`, `VersionTracking`, `Debugger`, `Emulator` . It is the gui you see after opening something in the project manager window.

I don't recommend using automated version tracking. You are likely to apply incorrect matches which will cascade to more incorrect matches and it'll just become a bad time.

**Comment by clearbluejar** [comment]
So `Tool` implies that it uses the GUI? Is there a way to around it?

I have seen some test environments such as https://github.com/NationalSecurityAgency/ghidra/blob/894bd3cfc2db78686727bbf2a1d369de92878033/Ghidra/Features/Base/src/main/java/ghidra/test/AbstractGhidraHeadlessIntegrationTest.java#L70-L74 that seem to fake out the tool and might get me there, although test isn't available in the standard release.

As for automated version tracking, in this case, I'm ok with incorrect matches. I'm just trying to take advantage of auto creating a VT session, and associating/applying matches the best it can.  

Hoping to be able to fake a tool enough to make the autoversiontrackingtask succeed:
https://github.com/NationalSecurityAgency/ghidra/blob/894bd3cfc2db78686727bbf2a1d369de92878033/Ghidra/Features/VersionTracking/src/main/java/ghidra/feature/vt/gui/actions/AutoVersionTrackingTask.java#L96

I don't require any interaction with the GUI, just need the session created and task run, and session saved. 

Is there a way?

**Reply by clearbluejar** [reply]
also.. just found this old thread.. "Getting a PluginTool in Headless Mode" https://github.com/NationalSecurityAgency/ghidra/issues/1944. Perhaps this issue is similar?

**Reply by dragonmacher** [reply, marked as answer]
> So Tool implies that it uses the GUI? Is there a way to around it?

Tool does explicitly mean a GUI is being used.   There is no way around that.  If you see code that uses a tool, then a GUI must be used.   This is a limitation that predates the headless API.

> Is there a way?

There is always a way, but we do not provide one.  

As @astrelsky pointed out, you should really be writing your headless scripts so that you can control what gets applied.  Doing so automatically can introduce mistakes.   I appreciate that the auto VT script is fairly safe.   Unfortunately, the way that script was written, it is wired to use the tool and other GUI concepts.  This was simply an oversight in the design.

It is possible for you to use the code that the auto VT script uses, but that will be quite advanced programming and is probably not worth it.   We do provide to VT scripts that should run in headless mode.  This scripts are meant to be used as examples to get people started.  They are:
- Ghidra/Features/VersionTracking/ghidra_scripts/OpenVersionTrackingSessionScript.java
- Ghidra/Features/VersionTracking/ghidra_scripts/FindChangedFunctionsScript.java

**Reply by clearbluejar** [reply]
> There is always a way,
> It is possible for you to use the code that the auto VT script uses, but that will be quite advanced programming and is probably not worth it. 

 I am going to dig around a bit and see what I can do.  

@astrelsky and @dragonmacher, thanks to both for clarity on the tool/GUI requirements. 

> Ghidra/Features/VersionTracking/ghidra_scripts/OpenVersionTrackingSessionScript.java
> Ghidra/Features/VersionTracking/ghidra_scripts/FindChangedFunctionsScript.java

These allow you to manipulate an already created session (even headless), and they are very helpful to get started, but my goal remains the ability to run the Auto VT task to create the session headless.

**Reply by astrelsky** [reply]
Usually for something like this I'll search the source code for some text from the gui to locate it. So I would go to create a version tracking session, find some notable text and track down the code that creates a new session and replicate it.

**Reply by dragonmacher** [reply]
> but my goal remains the ability to run the Auto VT task to create the session headless.

The work done by the auto vt action is here:
[AutoVersionTrackingTask](https://github.com/NationalSecurityAgency/ghidra/blob/e25881828cd240d39dc31e6d2430ebba4124918c/Ghidra/Features/VersionTracking/src/main/java/ghidra/feature/vt/gui/actions/AutoVersionTrackingTask.java)

The items that class uses that you will not have are the `serviceProvider`, the `controller` and any use of options.   The service provider can be faked by passing an instance of `ServiceProviderStub`.   That should work for the basic correlators.  you do not need options for your code, as you can hard-code any values you desire.   

The controller (which is a GUI concept) is used to get info about the matches and to get the current session.   You should be able to use the aforementioned scripts to figure out how to do that without the controller.

**Reply by clearbluejar** [reply]
Nice to know I can fake out `serviceProvider ` with `ServiceProviderStub`.  I'll try to update the discussion here as I make some progress.

**Reply by clearbluejar** [reply]
@dragonmacher , did this commit just open up headless VT?  https://github.com/NationalSecurityAgency/ghidra/commit/e7bc5efb931fbee84b1e2221de98954fe34ab2bb#diff-16fd51d62809bfd9e053d646b11a75f749b8391e4378bf3be3829aa11cf4314b

At a glance it seems like it did. I'll try it here soon. I had a version working in Python, but it involved creating some random awt swing threads and it essentially an abomination. This looks like a ton of work. Thank you.

**Reply by dragonmacher** [reply]
Yes, there were obstacles in the `AutoVerstionTrackingTask` that prevented its use in headless.  This commit removes those issues.   This task should be usable in headless now as-is.  As before, you can also use the other Version Tracking scripts mentioned in this discussion, along with any API that does not take a controller or a tool / service provider.

**Reply by clearbluejar** [reply]
This is amazing. thanks again. 

I have a new patch diff tool coming out here soon that does headless version tracking that could really benefit from this change. It should allow me to create VT sessions now, which has been a big feature I was hoping to provide.
